src/Kestrel.Core/Internal/HttpsConnectionMiddleware.cs

```diff
@@ -0,0 +1,291 @@
+                throw new ArgumentException(CoreStrings.ServerCertificateRequired, nameof(options));
+            }
+
+            // If a selector is provided then ignore the cert, it may be a default cert.
+            if (_serverCertificateSelector != null)
+            {
+                // SslStream doesn't allow both.
+                _serverCertificate = null;
+            }
+            else
+            {
+                EnsureCertificateIsAllowedForServerAuth(_serverCertificate);
+            }
+
+            _options = options;
+            _logger = loggerFactory?.CreateLogger(nameof(HttpsConnectionMiddleware));
+        }
+
+        public async Task OnConnectionAsync(ConnectionContext connectionContext)
+        {
+            SslStream sslStream;
+            bool certificateRequired;
+            var feature = new TlsConnectionFeature();
+            connectionContext.Features.Set<ITlsConnectionFeature>(feature);
+            connectionContext.Features.Set<ITlsHandshakeFeature>(feature);
+
+            var transportStream = new RawStream(connectionContext.Transport.Input, connectionContext.Transport.Output);
+
+            if (_options.ClientCertificateMode == ClientCertificateMode.NoCertificate)
+            {
+                sslStream = new SslStream(transportStream);
+                certificateRequired = false;
+            }
+            else
+            {
+                sslStream = new SslStream(transportStream,
+                    leaveInnerStreamOpen: false,
+                    userCertificateValidationCallback: (sender, certificate, chain, sslPolicyErrors) =>
+                    {
+                        if (certificate == null)
+                        {
+                            return _options.ClientCertificateMode != ClientCertificateMode.RequireCertificate;
+                        }
+
+                        if (_options.ClientCertificateValidation == null)
+                        {
+                            if (sslPolicyErrors != SslPolicyErrors.None)
+                            {
+                                return false;
+                            }
+                        }
+
+                        var certificate2 = ConvertToX509Certificate2(certificate);
+                        if (certificate2 == null)
+                        {
+                            return false;
+                        }
+
+                        if (_options.ClientCertificateValidation != null)
+                        {
+                            if (!_options.ClientCertificateValidation(certificate2, chain, sslPolicyErrors))
+                            {
+                                return false;
+                            }
+                        }
+
+                        return true;
+                    });
+
+                certificateRequired = true;
+            }
+
+            using (var cancellationTokeSource = new CancellationTokenSource(_options.HandshakeTimeout))
```

Comment:
This is no longer using `IConnectionTimeout` since it's added only added in HttpConnection.ProcessRequestAsync